source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+template <typename StatsStructType> class DirectStats;
+template <typename StatsStructType> class LazyInit;
+
+/**
+ * Interface for stats lazy initialization.
+ * To reduce memory and CPU consumption, Envoy can enable the bootstrap config
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>`.
+ * A 'StatsStructType' is only created when any of its field is referenced.
+ * See more context: https://github.com/envoyproxy/envoy/issues/23575
+ */
+template <typename StatsStructType> class LazyCompatibleInterface {
+public:
+  // Helper operators to get-or-create and return the StatsStructType object.
+  virtual StatsStructType* operator->() PURE;
+  virtual StatsStructType& operator*() PURE;
+  virtual ~LazyCompatibleInterface() = default;
+};
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ */
+template <typename StatsStructType>
+class LazyInit : public LazyCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::ScopeSharedPtr scope)
+      : inited_(Stats::Utility::gaugeFromElements(
+            *scope, {Stats::DynamicName(StatsStructType::typeName()), Stats::DynamicName("inited")},
+            Stats::Gauge::ImportMode::Accumulate)),
+        ctor_([scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
+          inited_.inc();
+          return new StatsStructType(stat_names, *scope);
+        }) {
+    if (inited_.value() > 0) {
+      instantiate();
+      ctor_ = nullptr;
+    }
+  }
+  // Helper operators to get-or-create and return the StatsStructType object.
+  inline StatsStructType* operator->() override { return instantiate(); }
```

Comment:
One simplification idea here. I think this class will never be directly accessed by anything, except for the new LazyCompatibleStats class, which makes everything a lot cleaner.

So my simplification here is modest. I think you don't need to use either operator overload here just make an instantiate() method that returns a reference, and you can just put the operator overloads on LazyCompatibleStats 
```
StatsStructType* operator->() { return &data_.instantiate(); }
StatsStructType* operator*() { return data_.instantiate(); }
```

It's a pretty modest improvement, but it means fewer virtual methods, fewer lines of code, and fewer levels of operator overloading.